From 1c496f3905eb0352b21344d390bd696c4430731d Mon Sep 17 00:00:00 2001
From: Neal Gompa <neal@gompa.dev>
Date: Wed, 25 Mar 2026 15:13:53 -0400
Subject: [PATCH] cmake: Refactor to use an EFI building module

This makes it straightforward to build EFI binaries using native
GCC.
---
 CMakeLists.txt              |  84 ++-------------
 cmake/Modules/FindEFI.cmake | 201 ++++++++++++++++++++++++++++++++++++
 2 files changed, 211 insertions(+), 74 deletions(-)
 create mode 100644 cmake/Modules/FindEFI.cmake

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 13b36b6..b3dd3a5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,45 +2,13 @@ cmake_minimum_required(VERSION 3.14)
 
 project(btrfs-efi LANGUAGES C VERSION 20230328)
 
-if(MSVC)
-    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /ENTRY:efi_main")
-    add_compile_options("/GS-")
-    string(REGEX REPLACE "/RTC(su|[1su])" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
-    string(REGEX REPLACE "/RTC(su|[1su])" "" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules")
 
-    # work around bug in Visual Studio
-    if (MSVC_CXX_ARCHITECTURE_ID STREQUAL "X86")
-        set(CMAKE_SYSTEM_PROCESSOR "X86")
-    endif()
-else()
-    add_compile_options(-fno-stack-check -fno-stack-protector -ffreestanding -Werror-implicit-function-declaration)
-endif()
+include(GNUInstallDirs)
 
-include_directories(${CMAKE_SOURCE_DIR}/gnu-efi/inc)
-if (CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-    include_directories(${CMAKE_SOURCE_DIR}/gnu-efi/inc/x86_64)
-elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "X86")
-    include_directories(${CMAKE_SOURCE_DIR}/gnu-efi/inc/ia32)
-elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "ARM64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
-    include_directories(${CMAKE_SOURCE_DIR}/gnu-efi/inc/aarch64)
-elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "ARM" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "arm")
-    include_directories(${CMAKE_SOURCE_DIR}/gnu-efi/inc/arm)
-endif()
+find_package(EFI REQUIRED)
 
-if(MSVC)
-    add_compile_options(-fpic -fpie -fshort-wchar -fno-ident)
-    add_compile_options(-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0)
-
-    if (CMAKE_SYSTEM_PROCESSOR STREQUAL "ARM64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
-        add_compile_options(-mcmodel=tiny)
-    elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-        add_compile_options(-mno-sse)
-    endif()
-else()
-    if (CMAKE_C_COMPILER_ID STREQUAL "Clang" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "X86")
-        add_compile_options(-mno-stack-arg-probe)
-    endif()
-endif()
+add_compile_options(-fno-stack-check -fno-stack-protector -ffreestanding -Werror-implicit-function-declaration)
 
 # -----------------------------------
 
@@ -56,12 +24,7 @@ set(ZSTD_SRC_FILES src/zstd/lib/common/entropy_common.c
 
 add_library(zstd STATIC ${ZSTD_SRC_FILES})
 target_compile_definitions(zstd PRIVATE -DZSTD_DEPS_MALLOC -DXXH_NO_STDLIB -DZSTD_HAVE_WEAK_SYMBOLS=0 -DZSTD_DISABLE_ASM)
-
-if(NOT MSVC)
-    target_compile_options(zstd PRIVATE -ffunction-sections -include ${CMAKE_SOURCE_DIR}/src/zstd-shim.h)
-else()
-    target_compile_options(zstd PRIVATE /Gy /FI ${CMAKE_SOURCE_DIR}/src/zstd-shim.h)
-endif()
+target_compile_options(zstd PRIVATE -ffunction-sections -include ${CMAKE_SOURCE_DIR}/src/zstd-shim.h)
 
 # -----------------------------------
 
@@ -75,12 +38,7 @@ set(ZLIB_SRC_FILES src/zlib/adler32.c
 
 add_library(zlib STATIC ${ZLIB_SRC_FILES})
 target_compile_definitions(zlib PRIVATE -DNO_GZIP -DZ_SOLO)
-
-if(NOT MSVC)
-    target_compile_options(zlib PRIVATE -ffunction-sections)
-else()
-    target_compile_options(zlib PRIVATE /Gy)
-endif()
+target_compile_options(zlib PRIVATE -ffunction-sections)
 
 # -----------------------------------
 
@@ -89,30 +47,8 @@ set(SRC_FILES src/btrfs.c
     src/misc.c
     src/lzo.c)
 
-add_executable(btrfs ${SRC_FILES})
-target_link_libraries(btrfs zstd zlib)
-
-if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
-    set_target_properties(btrfs PROPERTIES SUFFIX ".efi")
-else()
-    message(FATAL_ERROR "Please use mingw to compile")
-endif()
-
-if(${CMAKE_C_COMPILER_ID} STREQUAL "Clang")
-    target_link_options(btrfs PRIVATE "-Wl,--subsystem,efi_boot_service_driver")
-elseif(${CMAKE_C_COMPILER_ID} STREQUAL "GNU")
-    target_link_options(btrfs PRIVATE "-Wl,--subsystem,11")
-endif()
-
-if(${CMAKE_C_COMPILER_ID} STREQUAL "GNU" OR ${CMAKE_C_COMPILER_ID} STREQUAL "Clang")
-    target_link_options(btrfs PRIVATE -nostartfiles -nostdlib)
+add_library(btrfs STATIC ${SRC_FILES})
+target_link_libraries(btrfs zstd zlib efi)
 
-    if (${CMAKE_SYSTEM_PROCESSOR} STREQUAL "X86")
-        target_link_options(btrfs PRIVATE -e _efi_main)
-    else()
-        target_link_options(btrfs PRIVATE -e efi_main)
-    endif()
-elseif(MSVC)
-    target_link_options(btrfs PRIVATE "/SUBSYSTEM:EFI_BOOT_SERVICE_DRIVER")
-    target_compile_options(btrfs PRIVATE "/Oi-")
-endif()
+create_efi_image(btrfs btrfs)
+install_efi_image(btrfs ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
diff --git a/cmake/Modules/FindEFI.cmake b/cmake/Modules/FindEFI.cmake
new file mode 100644
index 0000000..678519a
--- /dev/null
+++ b/cmake/Modules/FindEFI.cmake
@@ -0,0 +1,201 @@
+# MIT License
+#
+# Copyright (c) 2018 Maxim Radugin <maxim.radugin@gmail.com>
+# Copyright (c) 2024 Neal Gompa <neal@gompa.dev>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+#[=======================================================================[.rst:
+FindEFI
+-------
+
+This module finds the gnuefi and efi libraries on the system.
+Additionally finds linker script and crt objects that corresponds to current
+platform (CMAKE_SYSTEM_PROCESSOR), provides create_efi_image() function
+to geneate resulting efi image.
+
+Add path where FindEFI.cmake is located to CMAKE_MODULE_PATH variable.
+For example, if FindEFI.cmake is located next to CMakeLists.txt file, use::
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR})
+
+CMAKE_MODULE_PATH should be set before calling find_package().
+See example project.
+
+For more information on find package modules see:
+https://cmake.org/cmake/help/latest/command/find_package.html
+
+The following will produce hello<arch>.efi image::
+
+  find_package(EFI REQUIRED)
+  add_library(hello_efi SHARED main.c)
+  target_link_libraries(hello_efi efi)
+  create_efi_image(hello_efi hello)
+  install_efi_image(hello_efi bin)
+
+The following variables are provided to indicate gnuefi support:
+
+.. variable:: EFI_FOUND
+
+  Variable indicating if the gnuefi and efi support was found.
+
+.. variable:: EFI_INCLUDE_DIRS
+
+  The directories containing the efi headers.
+
+.. variable:: EFI_C_FLAGS
+
+  The gnuefi and efi compilation flags
+
+.. variable:: EFI_LIBRARIES
+
+  The gnuefi and efi libraries to be linked.
+
+.. variable:: EFI_LINKER_FLAGS
+
+  The gnuefi and efi linker flags to be used when linking with these libraries.
+
+Additionally, the following :prop_tgt:`IMPORTED` target is being provided:
+
+.. variable:: efi
+
+  Interface target for using gnuefi and efi libraries. This target has all
+  the required libraries, flags and defintions set.
+
+.. variable:: efi::gnuefi
+
+  Imported target for gnuefi library (not to be used directly).
+
+.. variable:: efi::efi
+
+  Imported target for efi library (not to be used directly).
+
+The following cache variables may also be set:
+
+.. variable:: EFI_INCLUDE_DIR
+
+  The directory containing the efi headers.
+
+.. variable:: EFI_LIBRARY
+
+  The efi library.
+
+.. variable:: GNUEFI_LIBRARY
+
+  The gnuefi library.
+
+#]=======================================================================]
+
+find_library(EFI_LIBRARY NAMES efi DOC "efi library")
+
+find_path(EFI_INCLUDE_DIR efi/efi.h DOC "efi include directory")
+
+find_library(GNUEFI_LIBRARY NAMES gnuefi DOC "gnuefi library")
+
+# handle the QUIETLY and REQUIRED arguments and set xxx_FOUND to TRUE if
+# all listed variables are TRUE
+include(FindPackageHandleStandardArgs)
+
+find_package_handle_standard_args(efi
+    DEFAULT_MSG
+    EFI_LIBRARY
+    EFI_INCLUDE_DIR)
+
+find_package_handle_standard_args(gnuefi
+    DEFAULT_MSG
+    GNUEFI_LIBRARY)
+
+IF (EFI_FOUND AND GNUEFI_FOUND)
+    set(EFI_ARCH "${CMAKE_SYSTEM_PROCESSOR}")
+
+    if (EFI_ARCH MATCHES "amd64.*|x86_64.*|AMD64.*")
+        set(EFI_ARCH "x86_64")
+        set(EFI_OUT_ARCH "x64")
+    elseif (EFI_ARCH MATCHES "i686.*|i386.*")
+        set(EFI_ARCH "ia32")
+        set(EFI_OUT_ARCH "ia32")
+    elseif (EFI_ARCH MATCHES "aarch64.*|arm64.*|ARM64.*")
+        set(EFI_ARCH "aarch64")
+        set(EFI_OUT_ARCH "aa64")
+    elseif (EFI_ARCH MATCHES "armv.*|ARM")
+        set(EFI_ARCH "arm")
+        set(EFI_OUT_ARCH "arm")
+    elseif (EFI_ARCH MATCHES "riscv64.*")
+        set(EFI_ARCH "riscv64")
+        set(EFI_OUT_ARCH "riscv64")
+    else()
+        set(EFI_OUT_ARCH "")
+    endif()
+
+    get_filename_component(EFI_LIBRARY_DIR ${EFI_LIBRARY} DIRECTORY)
+    set(EFI_LDS "${EFI_LIBRARY_DIR}/elf_${EFI_ARCH}_efi.lds")
+    set(EFI_CRT_OBJS "${EFI_LIBRARY_DIR}/crt0-efi-${EFI_ARCH}.o")
+
+    set(EFI_BASE_C_FLAGS -ffreestanding -fno-stack-protector -fno-stack-check -fshort-wchar)
+    set(EFI_X86_C_FLAGS -mno-red-zone -maccumulate-outgoing-args -mno-stack-arg-probe)
+    if (EFI_ARCH MATCHES "x86_64.*|ia32.*")
+        set(EFI_C_FLAGS ${EFI_BASE_C_FLAGS} ${EFI_X86_C_FLAGS})
+    else()
+        set(EFI_C_FLAGS ${EFI_BASE_C_FLAGS})
+    endif()
+    set(EFI_LINKER_FLAGS ${EFI_CRT_OBJS} -Wl,-nostdlib -Wl,-znocombreloc -Wl,-T${EFI_LDS} -Wl,-Bsymbolic -Wl,--no-undefined)
+    set(EFI_INCLUDE_DIRS "${EFI_INCLUDE_DIR}/efi"
+        "${EFI_INCLUDE_DIR}/efi/protocol"
+        "${EFI_INCLUDE_DIR}/efi/${EFI_ARCH}")
+    set(EFI_LIBRARIES ${EFI_LIBRARY} ${GNUEFI_LIBRARY})
+
+    if(NOT TARGET efi::efi)
+        add_library(efi::efi IMPORTED STATIC)
+    endif()
+    set_property(TARGET efi::efi PROPERTY IMPORTED_LOCATION ${EFI_LIBRARY})
+
+    if(NOT TARGET efi::gnuefi)
+        add_library(efi::gnuefi IMPORTED STATIC)
+    endif()
+    set_property(TARGET efi::gnuefi PROPERTY IMPORTED_LOCATION ${GNUEFI_LIBRARY})
+
+    if(NOT TARGET efi)
+        add_library(efi INTERFACE)
+    endif()
+    set_property(TARGET efi PROPERTY INTERFACE_LINK_LIBRARIES ${EFI_LINKER_FLAGS} efi::efi efi::gnuefi)
+    set_property(TARGET efi PROPERTY INTERFACE_POSITION_INDEPENDENT_CODE ON)
+    set_property(TARGET efi PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${EFI_INCLUDE_DIRS})
+    set_property(TARGET efi PROPERTY INTERFACE_COMPILE_OPTIONS ${EFI_C_FLAGS})
+
+    mark_as_advanced(
+        EFI_LIBRARY
+        EFI_INCLUDE_DIR
+        GNUEFI_LIBRARY)
+endif()
+
+function (create_efi_image target out)
+    set(EFI_IMAGE_FILE ${CMAKE_CURRENT_BINARY_DIR}/${out}${EFI_OUT_ARCH}.efi)
+    add_custom_command(TARGET ${target} POST_BUILD
+        COMMENT "Creating EFI image: ${out}${EFI_OUT_ARCH}.efi"
+        COMMAND ${CMAKE_OBJCOPY} -j .text -j .sdata -j .data -j .dynamic
+                    -j .dynsym  -j .rel -j .rela -j .reloc
+                    --target=efi-app-${EFI_ARCH} --subsystem=11
+                    $<TARGET_FILE:${target}> ${EFI_IMAGE_FILE})
+
+    set_property(TARGET ${target} PROPERTY EFI_IMAGE_FILE ${EFI_IMAGE_FILE})
+endfunction()
+
+function (install_efi_image target dest)
+    install(FILES $<TARGET_PROPERTY:${target},EFI_IMAGE_FILE> DESTINATION ${dest})
+endfunction()
-- 
2.53.0

